cms/migrations/0033_placeholder_source_data_migration.py

```diff
@@ -0,0 +1,64 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11 on 2018-08-24 08:22
+from __future__ import unicode_literals
+
+from django.db import migrations
+
+
+def forwards(apps, schema_editor):
+    db_alias = schema_editor.connection.alias
+    ContentType = apps.get_model('contenttypes', 'ContentType')
+    CMSPlugin = apps.get_model('cms', 'CMSPlugin')
+    Placeholder = apps.get_model('cms', 'Placeholder')
+    filtered_content_types = (
+        ContentType
+        .objects
+        .using(db_alias)
+        .exclude(app_label='cms', model__in=['cmsplugin', 'placeholder'])
+    )
+
+    for ct in filtered_content_types:
+        model_class = apps.get_model(ct.app_label, ct.model)
+
+        if issubclass(model_class, (CMSPlugin, Placeholder)):
+            # We ignore all models inherited from CMSPlugin or Placeholder.
+            continue
+
+        # Get all fields which has a relationship to Placeholder.
+        pl_related_fields = [
+            f.name for f in model_class._meta.get_fields()
+            if f.related_model == Placeholder and not f.auto_created
+        ]
+
+        if not pl_related_fields:
+            # Model has no related fields with Placeholder.
+            continue
+
+        # Model has related field/s with Placeholder.
+        # Now get all the objects for the model and populate source field.
+        cur_ct_obj = ContentType.objects.get_for_model(model_class)
+
+        for obj in model_class.objects.using(db_alias):
+            for pl_field_name in pl_related_fields:
+                pl_field = getattr(obj, pl_field_name)
+
+                if pl_field.__class__.__name__ == 'ManyRelatedManager':
```

Comment:
use the collected fields above to avoid doing this check.
there's some methods on the field instances like get_queryset() which you might be able to leverage for an update call or an attribute to check for m2m